### PR TITLE
Feature/apl 1408 account control rest api

### PR DESCRIPTION
**Pull request by yuriy-larin-firstbridge**
Account control API methods

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+
+        CreateTransactionRequest txRequest;
+        try {
+            txRequest = HttpRequestToCreateTransactionRequestConverter.convert(
+                servletRequest, senderAccount, recipientAccount, recipientAccountId, 0, feeATM, attachment, broadcast);
+            log.trace("txRequest = {}", txRequest);
+        } catch(ParameterException e) {
+            log.warn("Tx conversion param exception", e);
+            return response.error(ApiErrors.CUSTOM_ERROR_MESSAGE, e.getErrorResponse()).build();
+        } catch(Exception e) {
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java:441]: Need to catch specific exception here, all other exceptions are treated as `unknown` and are being mapped by DefaultGlobalExceptionMapper

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+            log.warn("Tx global exception", e);
+            return response.error(ApiErrors.CUSTOM_ERROR_MESSAGE, "global exception: " + e.getMessage()).build();
+        }
+
+        Transaction transaction = txCreator.createTransactionThrowingException(txRequest);
+        log.trace("leaseBalance transaction = {}", transaction);
+        UnconfirmedTransactionDTO txDto = unconfirmedTransactionConverter.convert(transaction);
+        log.trace("leaseBalance txDto = {}", txDto);
+        LeaseBalanceResponse leaseBalanceResponse = new LeaseBalanceResponse();
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java:450]: Maybe move such operations to converter?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/parameter/AbstractRestParameter.java [hunk 3]
```diff
@@ -39,4 +39,13 @@ public T get() throws RestParameterException {
         }
         return value;
     }
+
+    @Override
+    public String toString() {
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/parameter/AbstractRestParameter.java:44]: Is default '@ToString' impl not sufficient?

File: apl-core/src/test/resources/conf/apl-blockchain.properties [hunk 4]
```diff
@@ -103,10 +103,10 @@ apl.testDbDir=./apl_test_db
 # Should unit test SQL data be loaded from db/data.sql ?
 apl.testData=true
 
-apl.dbParams=DB_CLOSE_ON_EXIT=FALSE
+apl.dbParams=DB_CLOSE_ON_EXIT=FALSE;AUTO_SERVER=TRUE
 
 # Deprecated,  testnet and mainnet dbs has the same configuration
-apl.testDbParams=DB_CLOSE_ON_EXIT=FALSE;MVCC=TRUE;MV_STORE=FALSE;TRACE_LEVEL_SYSTEM_OUT=3
+apl.testDbParams=DB_CLOSE_ON_EXIT=FALSE;MVCC=TRUE;MV_STORE=FALSE;TRACE_LEVEL_SYSTEM_OUT=3;AUTO_SERVER=TRUE
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/resources/conf/apl-blockchain.properties:109]: This option is deprecated, just remove it.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java (hunk 2 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java:450]: I will see if that json structure is used fir many methods... if that's true then it's worth to create converter.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java (hunk 1 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java:441]: Removed try-catch

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+        tags = {"accounts"})
+    @ApiResponse(description = "Transaction in json format", content = @Content(schema = @Schema(implementation = TransactionDTO.class)))
+    @PermitAll
+    @Secured2FA
+    public Response setPhasingOnlyControl(
+        @Parameter(required = true, schema = @Schema(implementation = String.class)) @Schema(description = "The sender account ID.")
+            @FormParam("sender") @NotNull AccountIdParameter senderAccountIdParameter,
+        @Parameter(required = true, schema = @Schema(implementation = VoteWeighting.VotingModel.class)) @Schema(description = "The expected voting model of the phasing. Possible values: NONE(-1), ACCOUNT(0), ATM(1), ASSET(2), CURRENCY(3)")
+            @FormParam("controlVotingModel") @DefaultValue("NONE") VoteWeighting.VotingModel controlVotingModel,
+        @Parameter(required = true, schema = @Schema(implementation = Long.class)) @Schema(description = "The expected quorum")
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java:176]: Second @Schema is used for correct displaying description on Swagger UI

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/converter/AccountControlPhasingConverter.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public AccountControlPhasingDTO convert(AccountControlPhasing model) {
+        return apply(model);
+    }
+
+    @Override
+    public List<AccountControlPhasingDTO> convert(List<AccountControlPhasing> models) {
+        return models.stream().map(this).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/converter/AccountControlPhasingConverter.java:51]: Removed

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+        Attachment attachment = new AccountControlEffectiveBalanceLeasing(period);
+
+        CreateTransactionRequest txRequest;
+        try {
+            txRequest = HttpRequestToCreateTransactionRequestConverter.convert(
+                servletRequest, senderAccount, recipientAccount, recipientAccountId, 0, feeATM, attachment, broadcast);
+            log.trace("txRequest = {}", txRequest);
+        } catch(ParameterException e) {
+            log.warn("Tx conversion param exception", e);
+            return response.error(ApiErrors.CUSTOM_ERROR_MESSAGE, e.getErrorResponse()).build();
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/AccountControlController.java:440]: Removing try-catch gives worse result on client side:
We have a error with HUGE stacktrace like:

{"protocol":"1","newErrorCode":2016,"requestProcessingTime":5, "errorDescription":"Unknown server error: '\nStrict stubbing argument mismatch. Please check:\n - this invocation of 'getAllStream' method:\n accountControlPhasingService.getAllStream(\n 0,\n 0\n);\n -> at com.apollocurrency.aplwallet.apl.core.rest.endpoint.AccountControlController.getAllPhasingOnlyControls(AccountControlController.java:138)\n - has following stubbing(s) with different arguments:\n 1. accountControlPhasingService.getAllStream(\n 0,\n 99\n);\n -> at com.apollocurrency...................... }

I don't think that is better for UI.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/parameter/AbstractRestParameter.java (hunk 3 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/parameter/AbstractRestParameter.java:44]: Updated

File: apl-core/src/test/resources/conf/apl-blockchain.properties (hunk 4 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/resources/conf/apl-blockchain.properties:109]: Removed

**Pull request merged**